Add "Forgot Password" Functionality

gaurav-jo1 opened the pull request:
**What does this PR do?**

This PR introduces the "Forgot Password" functionality to enhance user experience and security.

**What issues does this PR fix or reference?**

#411 

**If this PR changes the UI, include a screenshot below.**

https://github.com/user-attachments/assets/580a7dc7-f066-4f65-a7e2-d59b53cc6c01

**Additional Notes:**

* The `/forgot-password` endpoint response does not disclose whether the email exists in the system, preventing potential enumeration attacks.

* Added the `remove_existing_token_for_email` function to ensure that if a user makes multiple reset password requests, any previous tokens are invalidated.

Please let me know if there are any changes you’d like to make.

Winston-Hsiao commented on store/app/routers/users.py:
```diff
@@ -383,3 +383,62 @@ async def set_moderator(
+    try:
+        if user := await crud.get_user_from_email(data.email):
+            await crud.remove_existing_token_for_email(user.email)
+            reset_token = await crud.create_email_signup_token(email=user.email)
```
should have a new model/crud/routers for a `PasswordResetToken` model in `model.py` instead of re-using email sign up token as they serve different functions.

Winston-Hsiao commented on store/app/routers/users.py:
```diff
@@ -383,3 +383,62 @@ async def set_moderator(
+async def validate_password_reset_token(
+    data: ResetPasswordRequest, crud: Annotated[Crud, Depends(Crud.get)]
+) -> ResetPasswordResponse:
+    reset_token = await crud.get_email_signup_token(data.token)
```
again, let's utilize a new model here `PasswordResetToken`

Winston-Hsiao commented on store/app/routers/users.py:
```diff
@@ -383,3 +383,62 @@ async def set_moderator(
+    )
+
+    # Remove reset token
+    await crud.delete_email_signup_token(data.token)
```
Ensure all email signup token crud is changed to PasswordResetToken

Winston-Hsiao reviewed (changes requested):
Overall good, but some adjustments needed

gaurav-jo1 commented:
I have added the PasswordResetToken model and implemented new CRUD operations.

Please review the changes and let me know if any modifications or improvements are needed.

Winston-Hsiao reviewed (approved):
LGTM

Winston-Hsiao commented:
Hey @gaurav-jo1 sorry we've been busy shipping some larger features. There's some merge conflicts but we'd love to get your forgot password functionality you made merged in. Sorry for the delay!

gaurav-jo1 commented:
Hey Winston, no worries at all! I'd love to help get that merged. Since the UI has changed quite a bit recently, I'll need to update the UI in my PR as well. Give me a little time, and I'll get that sorted!
